### DISCUSSION TITLE
Spatio-temporal modeling of incompressible random vector field with 0 mean velocity

### DISCUSSION THREAD

**Discussion by joshua-benabou** [GeoStat-Framework/GSTools#245] [Q&A]
Hello, 

I am curious if GSTools has the ability to generate an incompressible 3D vector field of zero mean velocity, with a given correlation length in space, and a given correlation time-scale. I see that GSTools can do spatio-temporal modeling of random scalar fields as in this example 

https://geostat-framework.readthedocs.io/projects/gstools/en/stable/examples/09_spatio_temporal/02_precip_2d.html#sphx-glr-examples-09-spatio-temporal-02-precip-2d-py

which allows one to specify correlation scales in space and in time. I also see that generating random incompressible vector fields is also possible, as in this example:

https://geostat-framework.readthedocs.io/projects/gstools/en/stable/index.html#incompressible-vector-field-generation

To combine these capabilities, I encountered two issues:

1.  One is that the `gs.SRF` does not seem to have a setter method for the generator, so I am unsure how to modify the  generator `IncomprRandMeth` to my liking (to set 0 mean velocity in the x-direction instead of the default of 1). **Edit:** it appears passing `mean =(-1,0)` corrects the default x-velocity of 1 such that it now becomes 0. Still, I would like to know if the generator can be modified via a setter.

2. More importantly, adding e.g , `generator='VectorField'` in the first example linked above, simulates a sequence of vector fields which have 3D vectors at each point in (2+1)D space-time, instead of 2D vectors at each point in (2+1)D space-time (and the incompressibility condition applies to the (2+1)D vectors). I am interested in the second possibility, generalized to (3+1)D space-time, with the vector field incompressible at each time-step.  Is there a simple way around this? Here again I don't see how to modify the generator to do what I want? 

Forgetting about incompressibility, are there functions to generate a vector field other than 'IncomprRandMeth'?

**Edit:** looking a bit into the source code, it seems IncomprRandMeth could be modified to generate vectors with p components where p is not necessarily the dimensionality d of the specified covariance model, but some user-inputted parameter. For my case: p=3 and d = 3+1.

**Comment by LSchueler** [comment]
Hi joshua-benabou,

that's a really interesting application you're trying to set up there!

I don't have too much time today, but I'll try to get you on the right track and come back to you in a few days.

Here is a minimal working example of how you can access the generator (and all of its attributes, like the mean velocity in x-direction).

```python
import numpy as np
import gstools as gs
x = np.arange(100)
y = np.arange(100)
model = gs.Gaussian(dim=2, var=1, len_scale=10)
srf = gs.SRF(model, generator='VectorField', seed=19841203)
srf.generator.mean_u
```

Regarding your second problem, [take a look](https://github.com/GeoStat-Framework/GSTools/blob/5583a03031b91e1b66a0da766307e5b3b45f0a8b/src/gstools/field/generator.py#L331) at the way the `IncomprRandMeth` is created. In your own code, you can create a new class, also inherited by the `RandMeth` class and adjust it to your own liking.
... a few minutes pass by...
Okay, I think during a refactoring session something important for your use case got lost :sweat_smile:.  Normally you should be able to pass that new class to the `SRF` constructor, like so `srf = gs.SRF(model, generator=MyNewIncomprRandMeth, seed=383283)`
I will fix that *very* soon. In case this is very urgent, you could modify your local GSTools code in [this line](https://github.com/GeoStat-Framework/GSTools/blob/5583a03031b91e1b66a0da766307e5b3b45f0a8b/src/gstools/field/srf.py#L21) and add a lookup to your own class. You could probably also hotswap the generator and simply replace it with your own in an already created instance: `srf.generator = MyNewIncomprRandMeth(args)`

The incompressible field generator was actually the first one I created many years ago and at that time, I was only interested in 2d and 3d fields. Only recently have we started using spatio-temporal fields. I will have to look at the generator algorithm again to see, if it is easily expandable to higher dimensions. That would mean we would have to modify the heavy lifting Cython- and/or Rust code. I'm very interested in that and it would be great, if we could get that up and running!

As I already wrote, I will come back to you soon and I will fix the problem with not being able to use own generator classes in the `SRF` class.

**Comment by joshua-benabou** [comment]
Dear LSchueler,

thanks for your prompt reply. I am more than happy in the coming days to help extend the  incompressible field generator to spatio-temporal fields.

Based on my quick reading of the code, it seems the only special thing about the incompressible field generator is the application of a projector when summing the Fourier modes in the `gstools.field.summator`  function `summate_incompr`, to ensure incompressibility. For a time-varying incompressible field in 3 dimensions, this won't change. 

One thing is that the `_create_unit_vector` function in the source code for `gstools.field.generator` has a vector size equal to the dimension of the model. For a covariance model describing a random vector field evolving in 3 space + 1 time dimensions, we want the unit vector to have size 3, so perhaps there could be a flag for generating time-dependent vector fields, which sets the dimension of the vector field to the number of space dimensions, i.e (dimension of model) -1. But `_create_unit_vector` only matters for the mean velocity, so anywhere else the vector size appears, there would need to be changes.

The covariance matrix in the model might also have to be modified, I did not read that part of the code yet. For a regular NxNxN grid in 3D space, and N_t time points, there are random variables at 3*N^3 * N_t points, so that should be the column size of the covariance matrix. Otherwise, the correlation function is still separable between time and space. When naively setting the generator to the `IncomprRandMeth` in the  2 space + 1 time spatiotemporal field example I linked above, I was able to get a 3D vector field correlated in time; when then projecting to the two spatial dimensions, I got a 2D vector field correlated in time, but not incompressible. So I imagine somewhere in the code, the size of the covariance matrix is chosen based on whether we are using vectors or scalars, and if vectors, based on the length of the vector. 

It seems at a glance that the incompressibility and the covariance model are decoupled, so this could be as simple as changing the vector size, but I have not looked in detail yet.

**Reply by LSchueler** [reply]
Hi again,

first of all, I created the PR #250 to give better support for modified generators again. If you want to, please have a look at it and I'd be happy about comments.

> Based on my quick reading of the code, it seems the only special thing about the incompressible field generator is the application of a projector when summing the Fourier modes in the gstools.field.summator function summate_incompr, to ensure incompressibility. For a time-varying incompressible field in 3 dimensions, this won't change.

I think so too.

> So I imagine somewhere in the code, the size of the covariance matrix is chosen based on whether we are using vectors or scalars, and if vectors, based on the length of the vector.

Exactly, the relevant variable is `RandMeth._value_type`, which is either `"scalar"` or `"vector"`.

Just to summarize your ultimate goal for myself: You want to have a 4d field (3 spatial + 1 temporal) of incompressible 3d vectors, right? Thanks for already digging so deep into this topic and the relevant code. I think you are right, that this would not be too much of a problem.
I'll have a look at the equations and the code and come back to you soon.

**Comment by joshua-benabou** [comment]
Dear Lennart,

to confirm, yes I am looking to simulate a 4d field (3 space + 1 time) of
3d incompressible vectors, and this field should have a given correlation
length and correlation timescale.

thanks.

**Comment by LSchueler** [comment]
Hi Joshua,

have a look at PR #251. I'd be happy, if you would clone that branch and test it for your self and play around with it.
